Partially reverts backwards incompatible changes made in PR 13641

moorepants opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

Partially reverts #13641 

Fixes #14958 

#### Brief description of what is fixed or changed

    Reverts backwards incompatible changes to Matrix.diff()
    
    PR #13641 introduced a new feature: differentiating Matrices by Matrices
    and Arrays. To do this, three existing tests were changed, breaking
    backwards compatibility. The changes to the code caused
    MutableDenseMatrix.diff() to return an ImmutableDenseMatrix. This breaks
    user code that relies on the mutability of the matrices. In particular,
    it has large effects to the sympy.physics.mechanics code that relies
    heavily on matrix differentiation. This PR reverts the three changed
    tests, reverts the change to MatrixCalculus.diff() that causes the type
    change, and moves all subsequent tests related to differentiating
    Matrices with respect to Matrices/Array/etc into an XFAIL test, as they
    are now broken.


#### Other comments

This should be applied as a bugfix patch to SymPy 1.2 and 1.3.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* matrices
  * Reverts backwards incompatible change introduced in SymPy 1.2 and 1.3 for matrix differentiation and moves Matrix by Matrix differentiation to expected to fail.

<!-- END RELEASE NOTES -->

moorepants committed:
Reverted tests that were changed for #13641.

moorepants committed:
Reverts backwards incompatible changes to Matrix.diff()

PR #13641 introduced a new feature: differentiating Matrices by Matrices
and Arrays. To do this, three existing tests were changed, breaking
backwards compatibility. The changes to the code caused
MutableDenseMatrix.diff() to return an ImmutableDenseMatrix. This breaks
user code that relies on the mutability of the matrices. In particular,
it has large effects to the sympy.physics.mechanics code that relies
heavily on matrix differentiation. This PR reverts the three changed
tests, reverts the change to MatrixCalculus.diff() that causes the type
change, and moves all subsequent tests related to differentiating
Matrices with respect to Matrices/Array/etc into an XFAIL test, as they
are now broken.

moorepants committed:
Added test to ensure no type change in matrix differentiation.

Upabjojr commented:
> and moves all subsequent tests related to differentiating
Matrices with respect to Matrices/Array/etc into an XFAIL test, as they
are now broken.

I don't see the point in breaking tests, derivatives of matrices by matrices is an important feature.

What about just fixing the returned type of the matrix?

Upabjojr committed:
fixing issues for Jason Moore

moorepants commented:
> I don't see the point in breaking tests, derivatives of matrices by matrices is an important feature.

I don't see the point in breaking the physics tests in the first place. That screws up a decade of code written that utilizes the matrix module. This new feature is cool but no one using it yet. So I have no problem disabling it for those two reasons.

You self merged your own PR and modified existing tests. In my opinion, that is very poor practice.

> What about just fixing the returned type of the matrix?

You are welcome to fix everything properly. I don't have hours to figure it out.

>@moorepants I hope I've fixed the issues. I don't like to revert the changes, we really need to support matrix derivatives by other matrices, as this makes SymPy more powerful.

I don't like to revert them either, but it is necessary when poor choices were made and this becomes the necessary solution to maintain backwards compatibility.

smichr commented:
You can have it both ways by allowing `diff()` function to handle matrix by matrix derivation, but `.diff()` method to behave as it currently does.

moorepants commented on sympy/matrices/tests/test_commonmatrix.py:
```diff
@@ -1331,9 +1331,11 @@ def test_singular_values():
     x, y = symbols('x y')
     m = CalculusOnlyMatrix(2, 1, [x, y])
+    # TODO: currently not working as ``_MinimalMatrix`` cannot be sympified:
     assert m.diff(x) == Matrix(2, 1, [1, 0])
```
This should not be xfailed. It worked before your changes, no?

asmeurer commented:
Hi Jason. I've been busy this weekend with the mentor summit, though @Upabjojr was also there and we discussed this a little in person. 

I think we should try to fix this without breaking matrix wrt matrix differentiation if possible. I think it should be doable. The issue sounds like it just added a sympify somewhere, which converted MutableMatrix to ImmutableMatrix. I suspect this can be fixed by adding a check in `diff()`, as well as defining `MutableMatrix.diff` to only return mutable (assumedly it's less important if `Immutable.diff(Mutable)` gives mutable). We can figure out a more generic solution later, but hard-coding MutableMatrix for now is fine.

It would be helpful if you could test your upstream code with releases sooner, or even test against SymPy master. I also always do release candidates before releases and let them sit for at least a week. If this issue were caught before the final release was made I would have blocked the release to get it fixed. 

Regarding backporting releases, I had not planned on doing any of those. If you want you can make one (I can give you access). Everything is documented at https://github.com/sympy/sympy/blob/master/release/README.md. 

We could also do a 1.4 some time soon (we probably should before the end of the year). You can see the list of changes since 1.3 https://github.com/sympy/sympy/wiki/Release-Notes-for-1.4. Now that we have the release notes bot we can release more often, which I demonstrated by doing 1.3 a couple of months after 1.2. However, how often I can personally do releases in the future will depend on my situation, which is still somewhat up in the air.

moorepants commented:
> I think we should try to fix this without breaking matrix wrt matrix differentiation if possible.

I do to, but someone needs to put the time into doing that.  I have little interest in spending time to ensure the new matrix differentiation code works. I feel the burden of fixing this should go to @Upabjojr, especially since the original tests were changed in the PR that introduced this error, no deprecation warnings were added, and he self-merged it seemingly without signoff from others. It does not make me happy that I would have to spend time on this fix.

> It would be helpful if you could test your upstream code with releases sooner

I'd love to do that but in general my time is already spread thin (as is likely all of ours) and I don't see that as a reasonable request to all downstream packages that utilize SymPy. I don't have the luxury to do this at a whim. I find that if I leave a project stagnant for just 6 months that I almost always have to spend time fixing things when I come back to it because downstream scientific python projects do not remain backwards compatible. This is a big pet peeve of mine.

PyDy does test against SymPy master with every PR but if no one makes a PR in the SymPy release time window and I don't have time to do it, then it is what it is. I think the more reasonable expectation is that PRs to SymPy don't change tests and that SymPy devs try extremely hard to maintain backwards compatibility. If things have to absolutely be changed then we need deprecation warnings in one or more SymPy versions. It would also be relatively easy to add PyDy's (or any other upstream packages') tests to the release process of SymPy, which would IMHO be a better solution.

> Regarding backporting releases, I had not planned on doing any of those.

I know that SymPy does not have a history of doing these. I personally think that we should apply bugfix patches to old versions so that upstream packages can use specific SymPy versions reliably, but this is of course something that requires resources that we likely don't have. If I manage to find a day's worth of work to figure it all out then sure I'll do it, but that is unlikely to happen anytime soon.

Upabjojr commented:
>  I feel the burden of fixing this should go to @Upabjojr

I have already pushed my edit.

Upabjojr commented:
If nobody objects, I'm ok with meeting this PR.

Upabjojr commented:
> The CalculusOnlyMatrix test is still broken.

IIRC that class isn't used anywhere. We can open an issue for that. I won't be able to have a look until next week.

asmeurer commented:
Please add a test to make sure this doesn't regress again.

Upabjojr commented:
> Please add a test to make sure this doesn't regress again.

@moorepants has already added a test.

Upabjojr commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -1931,19 +1931,24 @@ def test_limit():
 
 def test_diff():
     A = MutableDenseMatrix(((1, 4, x), (y, 2, 4), (10, 5, x**2 + 1)))
+    assert isinstance(A.diff(x), type(A))
```
This test avoids future regressions.

Upabjojr committed:
Merge branch 'master' into revert-13641

asmeurer commented:
Is this ready? Ideally both @Upabjojr and @moorepants should sign off on this before it gets merged.

Upabjojr commented:
+1 to merge on my behalf.

Upabjojr commented:
Well, the support of derivatives of matrices by matrices is useful.

moorepants commented:
> Well, the support of derivatives of matrices by matrices is useful.

I completely agree, I've never said otherwise. What I don't agree with is changing tests. This state of this PR still leaves a test changed to XFAIL do the commits you added. I'm willing to let that one slide because that class may not be used publicly.

Upabjojr commented:
The class `_MinimalMatrix` does not inherit `MatrixBase`, therefore it can't be sympified. Furthermore, there's no way to call `.as_mutable()` on that class.

Maybe we should contact the author: @siefkenj .

Hi @siefkenj , we have an issue with the derivative of `_MinimalMatrix` now. The `.diff( )` method now sympifies the matrices and that class cannot apparently be sympified. As you appear to be the author of that class, do you have any suggestion?

asmeurer commented:
Why aren't any of the matrix classes in common.py subclassing from MatrixBase? Should the method be put on MatrixCommon or MatrixProperties instead?

moorepants commented:
I'd rather just merge this. It is better than it was.

asmeurer commented:
OK, we can open another issue about the XFAIL test.

siefkenj commented:
@asmeurer The basic idea is that common.py provides various matrix methods that are self contained, like arithmetic, or row reduction, etc. These could be cherrypicked for matrix-like objects. For instance, building a matrix over a ring, arithmetic makes sense, but row reduction or calculus may not. `MatrixBase` provides all methods common to both `DenseMatrix` and `SparseMatrix` (which means inheriting from all classes in common.py)

@Upabjojr `_MinimalMatrix` should not `sympify`. Because the matrix algorithms should apply generically, we shouldn't require inheritors to support `sympify`. (The background for this is that mutable matrices are not proper sympy objects, so we cannot make assumptions in the algorithms that we have proper sympy objects).

`CalculusOnlyMatrix` tests the calculus algorithms in isolation to make sure that you could cherry-pick them for a new type of matrix object and not run into any dependency issues. I haven't been paying attention lately, but I think the derivative wrt a matrix is implemented in a convoluted way. I would suggest adding a new class, say `MatrixTensorCalculus(MatrixCalculus)` that detects if the argument is a matrix-like object. If it is, it does the right thing. If not, it dispatches to `MatrixCalculus.diff`.

Glancing at the `Derivative` code, it seems it now imports things from the matrix class, but I think this could be avoided with appropriate definitions of `_eval_derivative`, etc.

Upabjojr commented:
> I would suggest adding a new class, say MatrixTensorCalculus(MatrixCalculus) that detects if the argument is a matrix-like object. If it is, it does the right thing. If not, it dispatches to MatrixCalculus.diff.

I'm a bit skeptical about creating too many classes, we risk ending up with code that is difficult to maintain.

> Glancing at the Derivative code, it seems it now imports things from the matrix class, but I think this could be avoided with appropriate definitions of _eval_derivative, etc.

The strangeness of `Derivative` is that you need to create a `Derivative` object even if you just want to calculate the evaluation of the derivative. The function `diff( )` calls `Derivative( ... ).doit()` in the end. This requires unfortunately that the objects to be derived to by sympified, which may not be optimal.

This issue could be solved by reviewing how `Derivative` works, but maybe there are some reasons why the code was written the way it is.

siefkenj commented:
Perhaps a better solution would be to add a `_eval_diff_wrt_matrix` method to `CalculusMatrix` and have `diff` dispatch depending on the types it's called with.

Upabjojr commented:
Are you ruling out the possibility for classes like `_MinimalMatrix` to have derivatives by matrices?

> Perhaps a better solution would be to add a _eval_diff_wrt_matrix method to CalculusMatrix and have diff dispatch depending on the types it's called with.

There is already a double dispatching mechanism to identify the method to use. The problem is that `.diff( )` currently passes through `Derivative`. `Derivative` then handles the dispatching to the proper function.

siefkenj commented:
@Upabjojr The original design was that `.diff()` called `_eval_diff()` directly.

The result of `Derivative(obj)` should always be a sympy object. The fact that `Derivative(mutable_mat)` now returns a `mutable_mat`, which is not a sympy object, is confusing. If `.diff()` calls `_eval_diff()` directly, this problem won't occur.

Upabjojr commented:
> If .diff() calls _eval_diff() directly, this problem won't occur.

Agreed. I have opened an issue https://github.com/sympy/sympy/issues/15435
